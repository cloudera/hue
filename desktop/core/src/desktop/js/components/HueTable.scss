/*
 Licensed to Cloudera, Inc. under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  Cloudera, Inc. licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

@import './styles/colors';
@import './styles/mixins';

.hue-table-container {
  overflow-x: auto;
  width: 100%;
  margin: 0 0 15px 0;

  .hue-table {
    line-height: 14px;
    table-layout: auto;
    border-collapse: separate;

    &.header-less {
      border-collapse: collapse;

      tbody > tr:first-child {
        border-top: 1px solid $fluid-gray-300;
      }
    }

    thead,
    tbody {
      tr {
        height: 40px;

        th,
        td {
          @include nowrap-ellipsis;

          height: 16px;
          max-width: 300px;
          padding: 12px;
          border-bottom: 1px solid $fluid-gray-300;
          text-align: left;

          &.column-flush {
            width: 100%;
          }
        }
      }
    }

    thead {
      tr {
        th {
          background-color: $fluid-white;
          color: $fluid-gray-700;
          font-size: 13px;
          font-weight: 500;
          white-space: nowrap;

          &.sticky-first-col {
            background-color: $fluid-white;
            position: sticky;
            position: -webkit-sticky;
            left: 0;
            z-index: 102;
          }

          .sort-header {
            display: flex;
            cursor: pointer;
            align-items: center;
            letter-spacing: normal;
            outline: 0;
          }
        }
      }
    }

    tbody {
      tr {
        td {
          color: $fluid-gray-900;
          font-size: 14px;

          &.sticky-first-col {
            @include position-sticky;

            background-color: $fluid-white;
            left: 0;
            z-index: 100;
          }

          &:last-of-type {
            padding-right: 8px;
          }
        }
      }

      &.clickable-rows tr {
        cursor: pointer;

        &:hover {
          background-color: $fluid-blue-050;
        }
      }
    }

    &.sticky-header {
      thead th {
        @include position-sticky;

        top: 0;
        z-index: 101;

        &.sticky-first-col {
          @include position-sticky;

          top: 0;
        }
      }
    }
  }
}
