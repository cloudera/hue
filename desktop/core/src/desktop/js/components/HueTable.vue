<!--
  Licensed to Cloudera, Inc. under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  Cloudera, Inc. licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

<!-- eslint-disable vue/no-v-html -->
<template>
  <div ref="tableContainer" class="hue-table-container" @scroll="onContainerScroll">
    <table
      class="hue-table"
      :class="{ 'sticky-header': stickyHeader && showHeader, 'header-less': !showHeader }"
    >
      <caption>
        <!-- Because of Web:TableWithoutCaptionCheck -->
        {{
          caption
        }}
      </caption>
      <thead v-if="showHeader">
        <tr class="header-row">
          <th
            v-for="(column, colIndex) in columns"
            :key="colIndex"
            :class="cellClass(column.headerCssClass, colIndex)"
            scope="col"
          >
            {{ typeof column.label !== 'undefined' ? column.label : column.key }}
          </th>
          <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
          <th class="column-flush" scope="col" />
        </tr>
      </thead>
      <tbody :class="{ 'clickable-rows': clickableRows }">
        <tr v-for="(row, rowIndex) in rows" :key="rowIndex" @click="onRowClick(row)">
          <td
            v-for="(column, colIndex) in columns"
            :key="colIndex"
            :class="cellClass(column.cssClass, colIndex)"
          >
            <slot v-if="hasCellSlot(column)" :name="cellSlotName(column)" v-bind="row" />
            <div v-else-if="column.htmlValue" v-html="row[column.key]" />
            <template v-else>
              {{ column.adapter ? column.adapter(column.key, row) : row[column.key] }}
            </template>
          </td>
          <td class="column-flush" />
        </tr>
      </tbody>
    </table>
  </div>
</template>
<!-- eslint-enable vue/no-v-html -->

<script lang="ts">
  import { defineComponent, PropType } from 'vue';

  import './HueTable.scss';
  import { Column, Row } from './HueTable';

  export default defineComponent({
    name: 'HueTable',
    props: {
      rows: {
        type: Array as PropType<Row[]>,
        required: false,
        default: () => []
      },
      columns: {
        type: Array as PropType<Column<unknown>[]>,
        required: false,
        default: () => []
      },
      caption: {
        type: String,
        default: undefined
      },
      showHeader: {
        type: Boolean,
        required: false,
        default: true
      },
      stickyHeader: {
        type: Boolean,
        required: false,
        default: false
      },
      stickyFirstColumn: {
        type: Boolean,
        required: false,
        default: false
      },
      clickableRows: {
        type: Boolean,
        default: false
      }
    },

    emits: ['scroll-to-end', 'row-clicked'],

    methods: {
      hasCellSlot(column: Column<unknown>): boolean {
        return !!this.$slots[this.cellSlotName(column)];
      },
      cellSlotName(column: Column<unknown>): string {
        return 'cell-' + column.key;
      },
      onContainerScroll(): void {
        const containerEl = <HTMLElement>this.$refs.tableContainer;
        if (containerEl.scrollHeight === containerEl.scrollTop + containerEl.clientHeight) {
          this.$emit('scroll-to-end');
        }
      },
      cellClass(cellClass: string | undefined, index: number): string | null {
        // This prevents rendering of empty class="" for :class="[x,y]" when x and y are undefined
        // Possibly fixed in Vue 3
        if (cellClass && this.stickyFirstColumn && index === 0) {
          return `${cellClass} sticky-first-col`;
        } else if (this.stickyFirstColumn && index === 0) {
          return 'sticky-first-col';
        }
        return cellClass || null;
      },
      onRowClick(row: Row): void {
        if (this.clickableRows) {
          this.$emit('row-clicked', row);
        }
      }
    }
  });
</script>
