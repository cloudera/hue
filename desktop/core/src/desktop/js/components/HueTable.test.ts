// Licensed to Cloudera, Inc. under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  Cloudera, Inc. licenses this file
// to you under the Apache License, Version 2.0 (the
// 'License'); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an 'AS IS' BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { mount, shallowMount } from '@vue/test-utils';
import { Column, Row } from './HueTable';
import HueTable from './HueTable.vue';

describe('HueTable.vue', () => {
  it('should render a table', () => {
    const wrapper = shallowMount(HueTable, {
      propsData: {
        columns: <Column<unknown>[]>[
          { key: 'a', label: 'A' },
          { key: 'd', label: 'D' },
          { key: 'c', label: 'C' },
          { key: 'b', label: 'B' }
        ],
        rows: <Row[]>[
          { a: '1', b: 5, c: false, d: undefined },
          { a: '2', b: 6, c: true, d: null },
          { a: '3', b: 7, c: false },
          { a: '4', b: 8, c: true }
        ]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a table with array rows', () => {
    const wrapper = shallowMount(HueTable, {
      propsData: {
        columns: <Column<unknown>[]>[
          { key: 1, label: 'A' },
          { key: 2, label: 'D' },
          { key: 3, label: 'C' },
          { key: 4, label: 'B' }
        ],
        rows: <Row[]>[
          ['1', 5, false, undefined],
          ['2', 6, true, null],
          ['3', 7, false],
          ['4', 8, true]
        ]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a table with a custom component for a cell', () => {
    const wrapper = mount(HueTable, {
      scopedSlots: {
        'cell-a': '<div>{{ props.a + props.b }}</div>'
      },
      propsData: {
        columns: <Column<unknown>[]>[
          { key: 'a', label: 'A' },
          { key: 'b', label: 'B' }
        ],
        rows: <Row[]>[
          { a: 1, b: 2 },
          { a: 4, b: 5 }
        ]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a table with adapter', () => {
    const wrapper = shallowMount(HueTable, {
      propsData: {
        columns: <Column<{ [key: string]: unknown }>[]>[
          { key: 'a', label: 'A' },
          {
            key: 'b',
            label: 'B + C',
            adapter: (key, row) => (row['b'] as number) + (row['c'] as number)
          }
        ],
        rows: <Row[]>[
          { a: '1', b: 5, c: 9 },
          { a: '2', b: 6, c: 10 },
          { a: '3', b: 7, c: 11 },
          { a: '4', b: 8, c: 12 }
        ]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render a sticky table with header and column css', () => {
    const wrapper = shallowMount(HueTable, {
      propsData: {
        columns: <Column<{ [key: string]: unknown }>[]>[
          {
            key: 'a',
            label: 'A',
            headerCssClass: 'header-css-class-A',
            cssClass: 'td-css-class-A'
          },
          { key: 'b', label: 'B', headerCssClass: 'header-css-class-B', cssClass: 'td-css-class-B' }
        ],
        stickyHeader: true,
        stickyFirstColumn: true,
        rows: <Row[]>[
          { a: '1', b: 5 },
          { a: '2', b: 6 },
          { a: '3', b: 7 },
          { a: '4', b: 8 }
        ]
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
