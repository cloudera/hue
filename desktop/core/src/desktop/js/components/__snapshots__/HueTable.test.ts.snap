// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HueTable.vue should render a sticky table with header and column css 1`] = `
<div
  class="hue-table-container"
>
  <table
    class="hue-table sticky-header"
  >
    <caption>
      <!-- Because of Web:TableWithoutCaptionCheck -->
      
    </caption>
    <thead>
      <tr
        class="header-row"
      >
        
        <th
          class="header-css-class-A sticky-first-col"
          scope="col"
        >
          A
        </th>
        <th
          class="header-css-class-B"
          scope="col"
        >
          B
        </th>
        
        <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
        <th
          class="column-flush"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      
      <tr>
        
        <td
          class="td-css-class-A sticky-first-col"
        >
          
          1
          
        </td>
        <td
          class="td-css-class-B"
        >
          
          5
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class="td-css-class-A sticky-first-col"
        >
          
          2
          
        </td>
        <td
          class="td-css-class-B"
        >
          
          6
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class="td-css-class-A sticky-first-col"
        >
          
          3
          
        </td>
        <td
          class="td-css-class-B"
        >
          
          7
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class="td-css-class-A sticky-first-col"
        >
          
          4
          
        </td>
        <td
          class="td-css-class-B"
        >
          
          8
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`HueTable.vue should render a table 1`] = `
<div
  class="hue-table-container"
>
  <table
    class="hue-table"
  >
    <caption>
      <!-- Because of Web:TableWithoutCaptionCheck -->
      
    </caption>
    <thead>
      <tr
        class="header-row"
      >
        
        <th
          class=""
          scope="col"
        >
          A
        </th>
        <th
          class=""
          scope="col"
        >
          D
        </th>
        <th
          class=""
          scope="col"
        >
          C
        </th>
        <th
          class=""
          scope="col"
        >
          B
        </th>
        
        <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
        <th
          class="column-flush"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      
      <tr>
        
        <td
          class=""
        >
          
          1
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          false
          
        </td>
        <td
          class=""
        >
          
          5
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          2
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          true
          
        </td>
        <td
          class=""
        >
          
          6
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          3
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          false
          
        </td>
        <td
          class=""
        >
          
          7
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          4
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          true
          
        </td>
        <td
          class=""
        >
          
          8
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`HueTable.vue should render a table with a custom component for a cell 1`] = `
<div
  class="hue-table-container"
>
  <table
    class="hue-table"
  >
    <caption>
      <!-- Because of Web:TableWithoutCaptionCheck -->
      
    </caption>
    <thead>
      <tr
        class="header-row"
      >
        
        <th
          class=""
          scope="col"
        >
          A
        </th>
        <th
          class=""
          scope="col"
        >
          B
        </th>
        
        <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
        <th
          class="column-flush"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      
      <tr>
        
        <td
          class=""
        >
          
          1
          
        </td>
        <td
          class=""
        >
          
          2
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          4
          
        </td>
        <td
          class=""
        >
          
          5
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`HueTable.vue should render a table with adapter 1`] = `
<div
  class="hue-table-container"
>
  <table
    class="hue-table"
  >
    <caption>
      <!-- Because of Web:TableWithoutCaptionCheck -->
      
    </caption>
    <thead>
      <tr
        class="header-row"
      >
        
        <th
          class=""
          scope="col"
        >
          A
        </th>
        <th
          class=""
          scope="col"
        >
          B + C
        </th>
        
        <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
        <th
          class="column-flush"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      
      <tr>
        
        <td
          class=""
        >
          
          1
          
        </td>
        <td
          class=""
        >
          
          14
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          2
          
        </td>
        <td
          class=""
        >
          
          16
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          3
          
        </td>
        <td
          class=""
        >
          
          18
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          4
          
        </td>
        <td
          class=""
        >
          
          20
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      
    </tbody>
  </table>
</div>
`;

exports[`HueTable.vue should render a table with array rows 1`] = `
<div
  class="hue-table-container"
>
  <table
    class="hue-table"
  >
    <caption>
      <!-- Because of Web:TableWithoutCaptionCheck -->
      
    </caption>
    <thead>
      <tr
        class="header-row"
      >
        
        <th
          class=""
          scope="col"
        >
          A
        </th>
        <th
          class=""
          scope="col"
        >
          D
        </th>
        <th
          class=""
          scope="col"
        >
          C
        </th>
        <th
          class=""
          scope="col"
        >
          B
        </th>
        
        <!-- To fill the blank space to the right when table width is smaller than available horizontal space -->
        <th
          class="column-flush"
          scope="col"
        />
      </tr>
    </thead>
    <tbody
      class=""
    >
      
      <tr>
        
        <td
          class=""
        >
          
          5
          
        </td>
        <td
          class=""
        >
          
          false
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          6
          
        </td>
        <td
          class=""
        >
          
          true
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          7
          
        </td>
        <td
          class=""
        >
          
          false
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      <tr>
        
        <td
          class=""
        >
          
          8
          
        </td>
        <td
          class=""
        >
          
          true
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        <td
          class=""
        >
          
          
          
        </td>
        
        <td
          class="column-flush"
        />
      </tr>
      
    </tbody>
  </table>
</div>
`;
